### ISSUE TITLE
github broken in 7.3.0

### ISSUE THREAD

**Issue by masterkain** [oauth2-proxy/oauth2-proxy#1724]
works in 7.2.1

 [oauthproxy.go:775] Error creating session during OAuth2 callback: unexpected status "404": {"message":"Not Found","documentation_url":"https://docs.github.com/rest/reference/users#list-email-addresses-for-the-authenticated-user"}

**Comment by rpardini** [comment]
indeed, 7.3.0 does not work but 7.2.1 does.
using `github_org` config, not team or user.
seems to be related to scopes: after downgrade, GH authorization page says "would like additional permissions to... Email addresses (read-only)" -- so possibly `user:email` scope is not being included when used with orgs in 7.3.0.

**Comment by multani** [comment]
I quickly checked this morning, and it seems to be coming from around 1f992b3f8704a210b003e374b7a2643b45caf386 or 82710a7ac167b745fa56667ca6848832c00ff4a1 (1f992b3f8704a210b003e374b7a2643b45caf386 doesn't build here).

The scope that was asked before was only `user:email`:

```
https://github.com/login/oauth/authorize?approval_prompt=force&client_id=XXX&redirect_uri=http://127.0.0.1:8000/oauth2/callback&response_type=code&scope=user:email&state=YYY
```

but now, OIDC-like scopes are requested instead (`openid email profile`):

```
https://github.com/login/oauth/authorize?approval_prompt=force&client_id=XXX&redirect_uri=http://127.0.0.1:8000/oauth2/callback&response_type=code&scope=openid email profile&state=YYY
```

**Comment by multani** [comment]
For now, it seems it can be fixed by explicitly configuring the `user:email` scope with either:

* the configuration flag `--scope="user:email"`
* the environment variable `OAUTH2_PROXY_SCOPE="user:email"`
* or the configuration file's setting `scope = "user:email"`

**Comment by JoelSpeed** [comment]
We broke the defaulting of provider configuration in 7.3.0 so a lot of options are needed to be set manually for now. Looking to fix this for the next release, though it's good practice to set them explicitly anyway so bugs like this don't break your config

**Comment by multani** [comment]
> though it's good practice to set them explicitly anyway so bugs like this don't break your config

It would be great to document the right settings somewhere then, I haven't try but I'm not sure the proxy works with a different set of scopes (or at least, if this one for GitHub `user:email` is missing) and it can be a bit hard to guess which values are supposed to be there (and I suppose they are different for every authentication providers...)

**Comment by BrynM** [comment]
Setting the scope to `user:email` as a workaround in the config did not work for a GH Enterprise (v3.5.5) install. I had to revert to 7.2.1 and remove that setting completely.

**Comment by kodeine** [comment]
scope to `user:email` did not work for me.

**Comment by jimeh** [comment]
I've restricted access to a specific team within a specific org. Setting scope to `user:email read:user read:org` worked for me.

**Comment by tuunit** [comment]
I've checked the current implementation and indeed the issue is described quite well in the #1903 bug ticket. Unfortunately, when not setting the scope to "user:email read:org" inside your config file, the default scope will be set to "openid email profile" which are not recognized by Github. Therefore the access does not work as your tokens will be created with insufficient permissions.

As of now, it is necessary to set the scope manually in your config file:
`scope = "user:email read:org"`

**Comment by tuunit** [comment]
PR with fix: https://github.com/oauth2-proxy/oauth2-proxy/pull/1926

**Comment by astronaut-chris** [comment]
Hello,

Is this resolved in 7.4.0?

**Comment by tuunit** [comment]
> Hello,
> 
> Is this resolved in 7.4.0?

No the fix will be part of the next release. Until then you will have to hardcode the default scope for GitHub.

Changelog:
https://github.com/oauth2-proxy/oauth2-proxy/blob/master/CHANGELOG.md

**Comment by chunjiw** [comment]
Neither `"user:email read:org"` nor `"user:email"` works for me.

**Comment by tuunit** [comment]
@chunjiw could you provide some more detail how you configure your environment and maybe share the code snippets of your configuration files?

**Comment by chunjiw** [comment]
@tuunit Sure!
oauth2-proxy is installed as a deployment in a kubernetes cluster, running in a container: 
```yml
        - args:
            - --provider=github
            - --email-domain=*
            - --cookie-domain=.int.company.com
            - --whitelist-domain=.int.company.com
            - --upstream=file:///dev/null
            - --http-address=0.0.0.0:4180
          image: quay.io/oauth2-proxy/oauth2-proxy:v7.2.1
```
The ingress to oauth2-proxy spec:
```yml
spec:
  rules:
    - host: "auth.int.company.com"
```
The ingress to my application:
```yml
metadata:
  name: app-name
  annotations:
    nginx.ingress.kubernetes.io/auth-url: "https://auth.int.company.com/oauth2/auth"
    nginx.ingress.kubernetes.io/auth-signin: "https://auth.int.company.com/oauth2/start?rd=https://app.int.company.com"
```

I largely followed a [Digital Ocean tutorial](https://www.digitalocean.com/community/tutorials/how-to-protect-private-kubernetes-services-behind-a-github-login-with-oauth2_proxy).

This is working, until I change to v7.3.0 or v7.4.0, even with `user:email read:org`

**Comment by tuunit** [comment]
@chunjiw the issue with v7.3.0 and v7.4.0
is that the scope is always overwritten with the default oidc scope if it has not been set via a config file. Therefore, if you want to use v7.3/4.0 you will have to mount a config map as file and use the config file instead of the arguments.

**Comment by chunjiw** [comment]
I see, thank you for your explanation! I will try it out later.

**Comment by chunjiw** [comment]
I confirm that adding `user:email` to config file instead of using args works without issue. Thank you!

**Comment by someburner** [comment]
Prefacing this comment: This is a great project and I appreciate all the work going into it! It's been so useful for me.

... but something like this should really *really* be noted somewhere more visible. I'm coming from a very old oauth2_proxy version and spent hours trying to figure out what was wrong with my config, hitting github rate limits due to this, etc, only to finally find this issue...

Just sayin' if there is a (rather major) problem with something for well over a year, maybe add a little blurb in the docs somewhere?

**Comment by batazor** [comment]
I understand this problem was solved in https://github.com/oauth2-proxy/oauth2-proxy/pull/1927
Maybe a new release of a new version should be released?

**Comment by tedder** [comment]
I think this is still a problem in the released version.

**Comment by jlarwig** [comment]
> > Hello,
> > Is this resolved in 7.4.0?
> 
> No the fix will be part of the next release. Until then you will have to hardcode the default scope for GitHub.
> 
> Changelog: https://github.com/oauth2-proxy/oauth2-proxy/blob/master/CHANGELOG.md

Hi @tedder,

as describe here, the fix is unfortunately not yet released and 7.4.0 is also broken.

**Comment by tuunit** [comment]
> > Hello,
> > Is this resolved in 7.4.0?
> 
> No the fix will be part of the next release. Until then you will have to hardcode the default scope for GitHub.
> 
> Changelog: https://github.com/oauth2-proxy/oauth2-proxy/blob/master/CHANGELOG.md

Hi @tedder,

as describe here, the fix is unfortunately not yet released and 7.4.0 is also broken.

The fix is not master but in no release as of now:

PR with fix: https://github.com/oauth2-proxy/oauth2-proxy/pull/1927

**Comment by BnJam** [comment]
I have created OAuth2 Proxy as a deployment

```yaml
containers:
 - image: quay.io/oauth2-proxy/oauth2-proxy:v7.5.0
```

Like everyone else, was struggling to get get it working and getting the same 404 Error, but for repositories, until I gave it the correct scope.

```yaml
env:
- name: OAUTH2_PROXY_SCOPE
  value: "read:user user:email repo"
```

Since we didn't want to add external users to our Org and only as collaborators to a repository - this worked for me on `v7.5.0` 

GH's Scope doc page was helpful - although I do with the `repo` scope was more restrictive. 
https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/scopes-for-oauth-apps 


As for my args, in case it's helpful: 

```yaml
args:
        - --http-address=0.0.0.0:4180
        - --upstream=file:///dev/null
        - --redirect-url=https://$host/oauth2/callback
        - --provider=github
        - --email-domain=*
        - --github-repo=OrgName/RepoName
        - --code-challenge-method=S256
        - --cookie-secure=true
        - --cookie-refresh=1h
        - --cookie-expire=24h
        - --cookie-name=proxy-cookie-name
        - --auth-logging=true
```

where `$host`, `OrgName/RepoName`, and `proxy-cookie-name` are filled in with my specific values (`$host` is my proxy address)

Hopefully this helps 🤞

**Comment by tuunit** [comment]
the issue was fixed with version 7.5.1

**Comment by tuunit** [comment]
> I have created OAuth2 Proxy as a deployment
> 
> ```yaml
> containers:
>  - image: quay.io/oauth2-proxy/oauth2-proxy:v7.5.0
> ```
> 
> Like everyone else, was struggling to get get it working and getting the same 404 Error, but for repositories, until I gave it the correct scope.
> 
> ```yaml
> env:
> - name: OAUTH2_PROXY_SCOPE
>   value: "read:user user:email repo"
> ```
> 
> Since we didn't want to add external users to our Org and only as collaborators to a repository - this worked for me on `v7.5.0`
> 
> GH's Scope doc page was helpful - although I do with the `repo` scope was more restrictive. https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/scopes-for-oauth-apps
> 
> As for my args, in case it's helpful:
> 
> ```yaml
> args:
>         - --http-address=0.0.0.0:4180
>         - --upstream=file:///dev/null
>         - --redirect-url=https://$host/oauth2/callback
>         - --provider=github
>         - --email-domain=*
>         - --github-repo=OrgName/RepoName
>         - --code-challenge-method=S256
>         - --cookie-secure=true
>         - --cookie-refresh=1h
>         - --cookie-expire=24h
>         - --cookie-name=proxy-cookie-name
>         - --auth-logging=true
> ```
> 
> where `$host`, `OrgName/RepoName`, and `proxy-cookie-name` are filled in with my specific values (`$host` is my proxy address)
> 
> Hopefully this helps 🤞

I'll have to check if we need to add additional logic to add the correct scope for repositories when using the repository restriction flag.
